src/fnft_nsev_slow.c

```diff
@@ -0,0 +1,683 @@
+            CHECK_RETCODE(ret_code, release_mem);
+        }
+        
+        ret_code = misc_merge(&K, buffer, SQRT(EPSILON));
+        CHECK_RETCODE(ret_code, release_mem);
+       
+        // Norming constants and/or residues)
+        if (normconsts_or_residues != NULL && *K_ptr != 0) {
+            
+            ret_code = compute_normconsts_or_residues(D, q, r, T, *K_ptr,
+                    bound_states, normconsts_or_residues, opts);
+            CHECK_RETCODE(ret_code, release_mem);
+            
+        }
+        *K_ptr = K;
+    } else if (K_ptr != NULL) {
+        *K_ptr = 0;
+    }
+    release_mem:
+        free(scatter_coeffs);
+        free(xi);
+        free(q_tmp);
+        return ret_code;
+}
+
+// Auxiliary function: Computes the norming constants and/or residues
+// using slow scattering schemes
+static inline INT compute_normconsts_or_residues(
+        const UINT D,
+        COMPLEX const * const q,
+        COMPLEX * r,
+        REAL const * const T,
+        const UINT K,
+        COMPLEX * const bound_states,
+        COMPLEX * const normconsts_or_residues,
+        fnft_nsev_slow_opts_t * const opts)
+{
+    
+    COMPLEX *a_vals = NULL, *aprime_vals = NULL;
+    UINT i, offset = 0;
+    INT ret_code = SUCCESS;
+    // Check inputs
+    if (K == 0) // no bound states to refine
+        return SUCCESS;
+    if (D == 0)
+        return E_INVALID_ARGUMENT(D);
+    if (q == NULL)
+        return E_INVALID_ARGUMENT(q);
+    
+    a_vals = malloc(K * sizeof(COMPLEX));
+    aprime_vals = malloc(K * sizeof(COMPLEX));
+    if (a_vals == NULL || aprime_vals == NULL) {
+        ret_code = E_NOMEM;
+        goto leave_fun;
+    }
+    
+    ret_code = nse_scatter_bound_states(D, q, r, T, K,
+            bound_states, a_vals, aprime_vals, normconsts_or_residues, opts->discretization, 0);
+    CHECK_RETCODE(ret_code, leave_fun);
+    
+    // Update to or add residues if requested
+    if (opts->discspec_type != nsev_dstype_NORMING_CONSTANTS) {
+        
+        if (opts->discspec_type == nsev_dstype_RESIDUES) {
+            offset = 0;
+        } else if (opts->discspec_type == nsev_dstype_BOTH) {
+            offset = K;
+            memcpy(normconsts_or_residues + offset,
+                    normconsts_or_residues,
+                    offset*sizeof(complex double));
+        } else
+            return E_INVALID_ARGUMENT(opts->discspec_type);
+        
+        // Divide norming constants by derivatives to get residues
+        for (i = 0; i < K; i++) {
+            if (aprime_vals[i] == 0.0)
+                return E_DIV_BY_ZERO;
+            normconsts_or_residues[offset + i] /= aprime_vals[i];
+        }
+    }
+    
+    leave_fun:
+        free(a_vals);
+        free(aprime_vals);
+        return ret_code;
+}
+
+
+// Auxiliary function for filtering: We assume that bound states must have
+// real part in the interval [-re_bound, re_bound].
+static inline REAL re_bound(const REAL eps_t)
```

Comment:
Code duplication, `re_bound` is already defined in `fnft_sev.c`. `im_bound` as well. The code for computing the norming constants is also almost identical and could be merged.

`fnft_nsev(...)` and `fnft_nsev_slow(...)` and `fnft_nsev_base(...)` and `fnft_nsev_slow_base(...)` also have almost identical parts that could potentially factored out, even though that looks at bit more involved. Thoughts?

Reply:
`re_bound`, `im_bound` and norming constants calculation could indeed be merged.  Although `fnft_nsev(...)` and `fnft_nsev_slow(...)` and `fnft_nsev_base(...)` and `fnft_nsev_slow_base(...)` have parts in common but there are significant enough differences that make me feel the code will get messy and difficult to read if merged.

Reply:
Ok, then I'd propose to do only the first three for now.

I'll have closer look at the rest on next. Maybe it would still make sense to factor at least some parts out.

Reply:
I thought about how to deal with the duplicated code. It seems the easiest solution would be not to have a second NFT function (`fnft_nsev_slow`), but to integrate the slow part into `fnft_nsev` instead.

One could determine from the discretization and bound_state_location parameters whether the transfer matrix has to be build in `fnft_nsev_base`. The functions `tf2contspec`, `tf2boundstates` and `tf2normconsts_and_residues` could be extended to deal with slow discretizations as well.

Thoughts?